src/dune_rules/pkg_rules.ml

```diff
@@ -414,51 +434,69 @@ module Substitute = struct
     let bimap t f g = { t with src = f t.src; dst = g t.dst }
     let is_useful_to ~memoize = memoize
 
-    let encode { vars; package; src; dst } input output : Dune_lang.t =
+    let encode { expander; src; dst } input output : Dune_lang.t =
       let e =
-        Package_variable.Map.to_list_map
-          vars
-          ~f:(fun { Package_variable.scope; name } v ->
-            let k =
-              let package =
-                match scope with
-                | Self -> Dune_sexp.List []
-                | Package p -> Dune_lang.Package_name.encode p
-              in
-              Dune_sexp.List [ package; Package_variable.Name.encode name ]
-            in
-            let v =
-              Dune_lang.atom_or_quoted_string (OpamVariable.string_of_variable_contents v)
-            in
-            Dune_sexp.List [ k; v ])
+        let paths (p : Paths.t) = p.source_dir, p.target_dir, p.name in
+        ( paths expander.paths
+        , expander.artifacts
+        , Package.Name.Map.to_list_map expander.deps ~f:(fun _ (m, p) -> m, paths p)
+        , expander.version
+        , expander.context
+        , expander.env )
+        |> Digest.generic
+        |> Digest.to_string_raw
+        |> Dune_sexp.atom_or_quoted_string
       in
-      let s = Dune_lang.Package_name.encode package in
-      List [ Dune_lang.atom_or_quoted_string name; List e; s; input src; output dst ]
+      List [ Dune_lang.atom_or_quoted_string name; e; input src; output dst ]
     ;;
 
-    let action { vars; package; src; dst } ~ectx:_ ~eenv:_ =
+    let action { expander; src; dst } ~ectx:_ ~eenv:_ =
       let open Fiber.O in
-      let+ () = Fiber.return () in
-      let env (var : Package_variable.t) =
-        match Package_variable.Map.find vars var with
-        | Some _ as v -> v
-        | None ->
-          Package_variable.Map.find
-            vars
-            { var with Package_variable.scope = Package package }
+      let* () = Fiber.return () in
+      let env (var : Substs.Variable.t) =
+        (* TODO loc *)
+        let loc = Loc.none in
+        let source =
+          (* TODO it's rather ugly that we're going through the pform machinery
+             to do this *)
+          { Dune_sexp.Template.Pform.loc; name = ""; payload = None }
+        in
+        let pform =
+          match var with
+          | Package var -> Package_variable.to_pform var
+          | Global n ->
+            let name = Package_variable.Name.to_string n in
+            (match Pform.Var.of_opam_global_variable_name name with
+             | Some v -> Var v
+             | None ->
+               User_error.raise ~loc [ Pp.textf "unrecognized variable name %S" name ])
```

Comment:
If I read the OPAM source code correctly, they resolve it to an empty string: https://github.com/ocaml/opam/blob/bb1400b92a7971cfd413aeb1c56076ac83b5b016/src/format/opamFilter.ml#L437

Reply:
Same as in shell scripts where you don't set `set -u` and have a typo in your variable.

But yeah if we have a way to print warnings to the user, resolving to an empty string and printing a warning to at least make people aware could be an option.